git server: add documentation and enable build start on push

csrwng opened the pull request:
- Updates the post-receive hook that will launch builds automatically.
- Removes extra main.go file
- Switches to use glog for logging and enables an environment variable for setting the log level
- Adds documentation on how to get started with the git server.

csrwng commented:
@bparees ptal

mfojtik commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+4. Update the PUBLIC_URL of your Git Server with the one provided by the route
+
+    ```sh
+    $ oc env dc/gitserver PUBLIC_URL=http://gitserver-myproject.infra.openshift.com
````
`oc set env` (oc env is deprecated, or will be soon).

mfojtik commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+
+   Edit the BuildConfig and set the source secret using the secret name returned above.
+   ```sh
+   $ oc edit bc ruby-hello-world
````
@smarterclayton "set the source secret" sounds like another thing we can add to `oc set`

smarterclayton replied:
Yeah, `oc set build-secret` or something.

kargakis replied:
`oc set secret` with subcommands - we will want to set secrets in a bunch of places

mfojtik commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+it a name of `helloworld` inside the Git Server.
+
+```sh
+$ oc env dc/gitserver GIT_INITIAL_CLONE_1=https://github.com/openshift/ruby-hello-world.git;helloworld
````
oc set env

mfojtik replied:
also not sure about the `;` character in bash

smarterclayton replied:
Probably should quote this.

bparees replied:
since you've got auto-deploy on config change, and an empty dir volume, users should be aware that this is going to destroy all their existing git repos.

mfojtik commented on examples/gitserver/hooks/buildconfig-autostart/post-receive:
```diff
@@ -0,0 +1,31 @@
+cat > ${commits}
+
+function startbuild { 
+  bcref=$1
```
nitpicking here... make this 'local' ;-)

mfojtik commented on examples/gitserver/hooks/buildconfig-autostart/post-receive:
```diff
@@ -0,0 +1,31 @@
+    # is included in the BuildConfig's reference
+    while read commit; do
+      if $(git merge-base --is-ancestor ${ref} ${commit}); then
+        echo "starting build for BuildConfig $1"
```
"Starting build for build configuration XYZ ..."

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	Reference  string
+}
+
+func (c *BuildConfigController) HandleBuildConfigChanges(deltas cache.Deltas) error {
```
godoc?

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	BCCache       map[string]BCCacheEntry
+}
+
+type BCCacheEntry struct {
```
godoc?

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+func (c *BuildConfigController) HandleBuildConfigChanges(deltas cache.Deltas) error {
+	// Do not handle any updates if we don't have a public URL
+	for _, d := range deltas {
+		bc := d.Object.(*buildapi.BuildConfig)
```
bc, ok := .... ? 
if !ok { error }

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	if !hasGitServerSource {
+		// If the BuildConfig's source doesn't point to this git server,
+		// ensure that any references to it are removed from existing git repositories.
+		entry, inCache := c.BCCache[bc.Name]
```
BCCache sounds weird ;-) BuildConfigCache ?

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	return nil
+}
+
+func (c *BuildConfigController) HandleBuildConfigDelete(bc *buildapi.BuildConfig) error {
```
godoc

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	if len(parts) > 1 {
+		return "", false
+	}
+	if !strings.HasSuffix(path, ".git") {
```
is this suffix needed?

csrwng replied:
It's added for consistency. You can create a repo ending with .git or without and from the client's perspective it's going to be the same thing. However, the server-side code could potentially create different dirs (one with and one without). This ensures that we always create the one with the suffix. We then create a symlink to the one without the suffix so they end up being the same thing.

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	dir := filepath.Join(c.Home, name)
+	info, err := os.Stat(dir)
+	if err != nil {
+		if os.IsNotExist(err) {
```
nuke this `if`

mfojtik replied:
(or provide fmt.Errorf() ;-)

csrwng replied:
Not sure I follow... so the if is there return a blank with no error if the directory simply doesn't exist

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+func (c *BuildConfigController) addToRepository(repo, bcref string) error {
+	glog.Infof("Adding BuildConfig reference %s to repository %s", bcref, repo)
```
shouldn't this be V(4)++ ?

csrwng replied:
yes

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	}
+	cmd := exec.Command("git", "config", "--add", "openshift.io.buildconfig", bcref)
+	cmd.Dir = dir
+	return cmd.Run()
```
in case this error, you will never see the error message, right? (CombinedOutput()) just something like "returned with code XYZ"

mfojtik replied:
but this will likely never error, so you can ignore my comment...

mfojtik replied:
(i prefer to have some helper/utility function to execute git command, capture output and in case of error, translate that output into a error message)

mfojtik commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+
+	if _, err := util.NewRepository(c.GitBinary, path, defaultHooks, &repoURL, nil); err != nil {
+		os.RemoveAll(path)
+		return "", fmt.Errorf("error: unable to initialize repo %s: %v", path, err)
```
do you need the `error:` prefix?

mfojtik commented on pkg/gitserver/controller/factory/factory.go:
```diff
@@ -0,0 +1,110 @@
+	bccontroller "github.com/openshift/origin/pkg/gitserver/controller"
+)
+
+type BuildConfigControllerFactory struct {
```
godoc

mfojtik commented on pkg/gitserver/gitserver.go:
```diff
@@ -184,12 +197,22 @@ func NewEnviromentConfig() (*Config, error) {
 		return nil, fmt.Errorf("only one of REQUIRE_SERVER_AUTH or REQUIRE_GIT_AUTH may be specified")
 	}
 
+	config.WatchBuildConfigs = os.Getenv("BUILDCONFIG_AUTOSTART") == "yes"
```
we better document the valid values ;-)

smarterclayton replied:
Why support this AND auto linking?  Why is this a boolean and not a "type"?  What other things would we want to add in the future?

bparees replied:
aren't the two orthogonal?  
auto-linking:  we figure out which buildconfig this repo belongs to.  
autostart=we run a build for the BC associated with this repo(if any) when a push happens.

autostart is unlikely to work without autolinking, but could if a user did the right configuration on their own.

i'm less clear on why watching buildconfigs is tied to autostart being true.  i would have thought it would be tied to autolinking.

mfojtik commented on pkg/gitserver/gitserver.go:
```diff
@@ -223,14 +247,15 @@ func NewEnviromentConfig() (*Config, error) {
 				}
 				return false, err
 			}
-			//log.Printf("debug: server response allowed=%t message=%s", res.Allowed, res.Reason)
```
nuke this ?

mfojtik commented on pkg/gitserver/initializer.go:
```diff
@@ -35,6 +34,9 @@ func lazyInitRepositoryHandler(config *Config, handler http.Handler) http.Handle
 			handler.ServeHTTP(w, r)
 			return
 		}
+		if !strings.HasSuffix(name, ".git") {
```
same comment as above... is the suffix really required?

csrwng replied:
see explanation above

mfojtik commented on pkg/gitserver/util/hooks.go:
```diff
@@ -0,0 +1,30 @@
+	"github.com/golang/glog"
+)
+
+func LoadHooks(path string) (map[string]string, error) {
```
godoc

mfojtik commented on pkg/gitserver/util/repository.go:
```diff
@@ -0,0 +1,78 @@
+	"github.com/openshift/origin/pkg/generate/git"
+)
+
+func NewRepository(gitBinary, path string, hooks map[string]string, self *url.URL, origin *url.URL) ([]byte, error) {
```
godoc

smarterclayton commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+2. Grant `view` access to the `gitserver` service account
+
+    ```sh
+    $ oc policy add-role-touser view -z gitserver
````
`add-role-to-user`

smarterclayton commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+    $ oc create -f gitserver.yaml
+    ```
+
+2. Grant `view` access to the `gitserver` service account
````
I think we should call it the `git` service account and the service `git` and the route `git`.  And probably the dc and a bunch of other things.

bparees replied:
+1

smarterclayton commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+2. Clone a new repository
+   
+   ```
+   $ git clone http://gitserver-myproject.infra.openshift.com/ruby-hello-world.git
````
Why do I have to clone first?  99% of the time I have a git project locally?

csrwng replied:
No reason... it's just one possible flow. If pushing first is more common, I'll use that.

bparees replied:
having a local repo and pushing to the git server definitely seems more likely to me.

i'd do clone from the github repo, then push to the git server.

smarterclayton replied:
```
git clone github.com/foo/bar.git
git remote add openshift ...
git push openshift
```

smarterclayton commented on examples/gitserver/gitserver.yaml:
```diff
@@ -138,6 +138,20 @@ items:
+          # If configured, and using the buildconfig-autostart post-receive
+          # hook, the git server will automatically start builds when commits
+          # are pushed to repositories for which BuildConfigs exist. This 
+          # requires that the 
```
Your comment needs tone updated.

smarterclayton commented on examples/gitserver/hooks/buildconfig-autostart/post-receive:
```diff
@@ -0,0 +1,31 @@
+}
+
+export -f startbuild
+git config --get-all openshift.io.buildconfig | xargs -n 1 -I {} bash -c 'startbuild "$@"' _  {}
```
Some comments here would be nice.

smarterclayton commented on examples/gitserver/hooks/buildconfig-autostart/post-receive:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Describe this script and what it is supposed to do.

kargakis commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+3. Create a route to the Git Server
+
+    ```sh
+    $ oc expose service/gitserver
````
Since we recommend using secure routes, we should also use them in the example.

kargakis commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,220 @@
+   just modify the route to use edge termination:
+
+   ```sh
+   $ oc edit route gitserver
````
Use `oc create route` and have the server secured in the first place?

csrwng replied:
@kargakis why do we have `oc create route` and `oc expose`? Are we deprecating one of them? I didn't see an option in `oc expose` to set edge termination.

kargakis replied:
Right, we are not going to enhance `oc expose` for creating secured routes since the command is already bloated from upstream, that's why we spinned off `create route`.

kargakis commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+)
+
+// BuildConfigController watches build resources and manages their state
+type BuildConfigController struct {
```
@deads2k fyi, another cache we will want to share

csrwng replied:
@kargakis this is not part of the main process. It's created in every git server you launch.

smarterclayton commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	Home          string
+	GitBinary     string
+	Client        osclient.Interface
+	BCCache       map[string]BCCacheEntry
```
Why are you using your own cache when we already have caches for this?

smarterclayton commented on pkg/gitserver/controller/factory/factory.go:
```diff
@@ -0,0 +1,110 @@
+// Create creates a new ConfigChangeController which is used to trigger builds on creation
+func (f *BuildConfigControllerFactory) Create() controller.RunnableController {
+	queue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, &keyListerGetter{})
+	cache.NewReflector(&buildConfigLW{client: f.Client, namespace: f.Namespace}, &buildapi.BuildConfig{}, queue, 2*time.Minute).RunUntil(f.Stop)
```
Please don't use 2 minute sync intervals.

smarterclayton commented on pkg/gitserver/controller/factory/factory.go:
```diff
@@ -0,0 +1,110 @@
+
+// Create creates a new ConfigChangeController which is used to trigger builds on creation
+func (f *BuildConfigControllerFactory) Create() controller.RunnableController {
+	queue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, &keyListerGetter{})
```
If you're using a DeltaFIFO you should be providing the list of the things that the git controller knows about, otherwise there's no cleanup that happens on your side.

smarterclayton commented on pkg/gitserver/controller/controller.go:
```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return "", false
+	}
+	if parsedURL.Host != c.PublicURL.Host ||
```
The fragility of knowing whether a build config points at you is the reason why I got blocked on the gitserver auto linking - because the git url could be any one of a hundred variations that doesn't quite match.  The simplest misconfiguration would totally break this logic silently, which means hard to debug issues for end users.

I think you need a stronger signal here.  I think there are three flows that are relevant:

1. User sets up gitserver, pushes, and hopes it works.  I would *expect* pushing to trigger the build with the same name as the repo via binary build.  That's straight up simpler than this flow.
2. User does the more complex "point build config at local git server because they don't have their own repo" - that's a detection flow, but when you detect you should record a stronger signal than just "the current URL happens to match", like an annotation
3. User wants to have their build config point at a remote, but mirror and do polling.  In this case you're going to need more than this check.

smarterclayton replied:
Why do you even need to have this controller?  Why not just do the check on push?

csrwng replied:
What is the use case for a binary push? (vs invoking the web hook)

bparees replied:
one advantage to *not* doing binary push is that you can rerun the build.

smarterclayton commented on pkg/gitserver/util/hooks.go:
```diff
@@ -0,0 +1,30 @@
+package util
```
Don't create util packages.  Create a package named for what this package is going to accomplish, like `pkg/gitserver/hooks`.

bparees commented on examples/gitserver/gitserver.yaml:
```diff
@@ -138,6 +138,20 @@ items:
+          # hook, the git server will automatically start builds when commits
+          # are pushed to repositories for which BuildConfigs exist. This 
+          # requires that the 
+          - name: BUILDCONFIG_AUTOSTART
```
i think this name could be better.  BUILDCONFIG_START_ON_PUSH?  BUILD_ON_PUSH?

bparees commented on examples/gitserver/gitserver.yaml:
```diff
@@ -138,6 +138,20 @@ items:
+
+          # The namespace to use to watch/launch BuildConfigs that point to 
+          # repositories from  this server.
+          - name: BUILCONFIG_NAMESPACE
```
missing a D in the name?

bparees commented on examples/gitserver/hooks/buildconfig-autostart/post-receive:
```diff
@@ -0,0 +1,31 @@
+  bc=${bcref[0]}
+  ref=${bcref[1]:-master}
+  
+  if webhook=$(oc start-build --list-webhooks="generic" "${bc}" | head -n 1); then
```
i don't get why you care if the buildconfig has a webhook or why you're using the webhook to start the build instead of just oc start-build?

bparees commented on examples/gitserver/hooks/code-detection/post-receive:
```diff
@@ -13,7 +13,7 @@ function addkey {
 function detect {
   if detected=$(key openshift.io.detect); then
-    exit 0
+    return 0
```
this entire hook is not currently used/referenced right?

bparees commented on pkg/gitserver/controller/factory/factory.go:
```diff
@@ -0,0 +1,110 @@
+	Stop <-chan struct{}
+}
+
+// Create creates a new ConfigChangeController which is used to trigger builds on creation
```
BuildConfigController

bparees replied:
rest of godoc is wrong also

bparees commented on pkg/gitserver/gitserver.go:
```diff
@@ -60,6 +62,14 @@ GIT_INITIAL_CLONE_*=<url>[;<name>]
   process will halt. If the repository already exists on disk, it will be updated from the remote.
+BUILDCONFIG_AUTOSTART=yes|no
+  if 'yes', builds will be started automatically for BuildConfigs that have a generic webhook defined
+  and are configured with source on this server.
```
1) why do we care if it has a webhook?
2) build will be started automatically *on push* for BuildConfigs that......    right?

bparees replied:
(still don't really like the name.  BUILD_ON_PUSH seems better)

bparees replied:
also true/false is more standard i think (though i admit less "englishy" in this case)

csrwng commented:
Updated pull - now using the name of the repository to kick off builds, removed controller.

bparees commented on examples/gitserver/README.md:
```diff
@@ -1,16 +1,207 @@
+   By pushing to it, you are creating it in the Git Server.
+
+   On push the git server will invoke new-app on the code and create artifacts for it in 
+   OpenShift.
```
so if i don't want new-app invocation because i've already setup my buildconfig/deploymentconfig, how do i get my repo created w/o this happening?

csrwng replied:
The new-app invocation won't happen because the bc will be found and a build will be started.

bparees commented on examples/gitserver/README.md:
```diff
@@ -1,16 +1,207 @@
+### Secure the Git Server
+
+1. If your cluster's router has been setup with a default certificate for edge termination,
+   just modify the route to use edge termination:
```
why don't we default the route to edge termination?

csrwng replied:
to simplify initial setup - if we're saying that the greatest usefulness of the git server right now is to quickly get a demo going, having to deal with self-signed ca certs will get in the way. Of course you'd never run in production that way, but we're not saying this is for production (yet).

bparees replied:
so what's the default?  insecure route?

csrwng replied:
yes

bparees commented on examples/gitserver/README.md:
```diff
@@ -1,16 +1,207 @@
+
+   **NOTE:** changing environment variables on the git server will cause a redeploy of the git server. If using
+   the default ephemeral storage for it, all repositories that have been pushed previously will be wiped out.
+   They will need to be pushed to the server again to restore.
```
which again will invoke new-app...so is there some check that determines if a BuildConfig already exists for the repo before it invokes new-app?

csrwng replied:
Yes. The build config will be found and its build invoked, new-app won't be called:
https://github.com/csrwng/origin/blob/gitserver_bc_controller/examples/gitserver/hooks/post-receive#L77

bparees commented on examples/gitserver/README.md:
```diff
@@ -1,16 +1,207 @@
+the credentials are the user name, and the user's token (from `oc whoami -t`). For a service account, it can
+be one of the 2 tokens created with the service account and stored in the service account secrets. These can 
+be obtained by looking at the secrets that correspond to the service account (`oc get secrets`) and displaying
+the token in one of them: `oc describe secret secret-name`.
```
what do i do w/ this token?  how do i tell the gitserver whether to use a user account or a service account, and which service account?

csrwng replied:
I'll clarify that the service account is any service account that can create pods, and the username can be the service account name (although it really doesn't matter, but it has to be 'something'). And the password is the service account token.

bparees commented on examples/gitserver/README.md:
```diff
@@ -1,16 +1,207 @@
+-----------------------------
+
+By default, whenever the git server receives a commit, it will look for a BuildConfig named the same as the
+repository where the commit is being pushed. If it finds a BuildConfig with the same name, it will start a build
```
in the same namespace.

csrwng replied:
thx will update

bparees commented on examples/gitserver/README.md:
```diff
@@ -1,16 +1,207 @@
+By default, whenever the git server receives a commit, it will look for a BuildConfig named the same as the
+repository where the commit is being pushed. If it finds a BuildConfig with the same name, it will start a build
+for that BuildConfig. Alternatively, an annotation may be added to a BuildConfig to explicitly link it to a 
+repository on the git server:
```
will it start builds for all matching buildconfigs, or just the first one?  (suppose i add the annotation to multiple BCs)

csrwng replied:
for all matching build configs
https://github.com/csrwng/origin/blob/gitserver_bc_controller/examples/gitserver/hooks/post-receive#L58-L59

bparees commented on examples/gitserver/README.md:
````diff
@@ -1,16 +1,207 @@
+   $ git config --global http.https://gitserver-myproject.infra.openshift.com.sslCAInfo /path/to/ca.crt
+   ```
+
+   where the key is http.[git server URL].sslCAInfo
````
I have to do this *and* modify the route to use edge termination if i'm using a private CA, right?  it wasn't totally clear to me if (1) and (2) were "either/or" or "(1) is mandatory and (2) is optional"

csrwng replied:
Will clarify

csrwng replied:
hmm... so you may not have a private CA... in which case you don't need step 2

bparees commented on examples/gitserver/hooks/post-receive:
```diff
@@ -7,48 +7,74 @@ set -o pipefail
+  # The purpose of using a custom detection script is that users can customize the 
+  # 'detect-language' script to return the image that's appropriate for their needs. 
+  # It is possible to just have new-app do code detection. In that case, just set
+  # the customDetection variable to false
```
i'm not convinced we should have this enabled by default since it's going to always match maven projects to wildfly, whereas new-app would match them to either wildfly or EAP depending on which imagestreams the cluster has available.

bparees replied:
and what customDetection variable?  the DETECTION_SCRIPT variable?  (which would need to be set to empty string, not false)

csrwng replied:
Will fix

bparees commented:
mostly just some suggestions to expand the doc and one concern about the language detect behaivor. otherwise lgtm.

csrwng commented:
@bparees made changes to address your comments, ptal

bparees commented on examples/gitserver/hooks/post-receive:
```diff
@@ -7,48 +7,77 @@ set -o pipefail
-    exit 0
+  local repoURL="$1"
+  local repoName="$2"
+  local detectionScript=${DETECTION_SCRIPT:-}
```
does this variable serve any purpose anymore?

csrwng replied:
yes, if you set it to 'detect-language' then it will use the script

csrwng replied:
or whatever other script name you come up with

bparees replied:
i mean you don't need the defaulting behavior.  you could just reference $DETECTION_SCRIPT everywhere, because the defaulting behavior here is saying "if $DETECTION_SCRIPT is set to nothing, set detectionScript to nothing"

csrwng replied:
If $DETECTION_SCRIPT is not set and I try to reference it, I believe my script will break

bparees replied:
ah, you have nounset.
ok.

bparees commented on pkg/gitserver/gitserver.go:
```diff
@@ -50,16 +52,28 @@ ALLOW_LAZY_CREATE
   process will halt. If the repository already exists on disk, it will be updated from the remote.
+AUTOLINK_KUBECONFIG
+  The location to read auth configuration from for autolinking.
+  If '-', use the service account token to link. The account
```
literally a "-" value?  or empty value?

csrwng replied:
It is a "-"
https://github.com/csrwng/origin/blob/gitserver_bc_controller/pkg/gitserver/autobuild/autobuild.go#L76

bparees replied:
cute.

bparees commented on examples/gitserver/gitserver.yaml:
```diff
@@ -70,12 +70,17 @@ items:
+          # If 'yes' new-app will be invoked on push for repositories
+          # for which a matching BuildConfig is not found.
+          - name: GENERATE_ARTIFACTS
+            value: "yes"
```
still bothers me we are using "yes" for this when we use "true" everywhere else i can think of.

csrwng replied:
I can change it. I'm just being consistent with what we already had.

bparees replied:
unless @smarterclayton wants to defend the decision i'd like to change it.

smarterclayton replied:
I don't have an issue with changing these flags to whatever is consistent.

On Tue, Apr 19, 2016 at 6:21 PM, Ben Parees <notifications@github.com>
wrote:

> In examples/gitserver/gitserver.yaml
> <https://github.com/openshift/origin/pull/8394#discussion_r60322525>:
>
> > @@ -70,12 +70,17 @@ items:
> >            - name: HOOK_PATH
> >              value: /var/lib/git-hooks
> >
> > -          # The script to use to do custom language detection on a
> > +          # If 'yes' new-app will be invoked on push for repositories
> > +          # for which a matching BuildConfig is not found.
> > +          - name: GENERATE_ARTIFACTS
> > +            value: "yes"
>
> unless @smarterclayton <https://github.com/smarterclayton> wants to
> defend the decision i'd like to change it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8394/files/b753c78d158936df66b3ee42e706145414da118b..1b2dc0730486fb06774d4b77e49e4b3990d007d2#r60322525>
>

bparees commented:
resolve the yes/true debate and then lgtm.

bparees commented:
(squash implied)

csrwng commented:
@bparees In addition to review changes I've added a couple of commits to resolve issues I ran into while testing. The anonymous pull wasn't working when the Authorization header was empty. It also looks like routing to a service doesn't work within a pod if the same pod is the destination of the service. I added an alias in the gitserver's gitconfig so that calling new-app http://git:8080/blah would work correctly.

bparees commented:
@csrwng new commits lgtm.

csrwng commented:
@bparees is this ok to merge?

bparees commented:
[merge]

hallelujah :)

csrwng commented:
[merge]

csrwng commented:
[merge]

csrwng commented:
[merge]
